pulpcore/cli/rpm/repository.py

```diff
@@ -159,7 +159,11 @@ def repository(ctx: click.Context, pulp_ctx: PulpCLIContext, repo_type: str) ->
             on the repodata."""
         ),
     ),
-    click.option("--sqlite-metadata/--no-sqlite-metadata", default=None),
+    pulp_option(
+        "--sqlite-metadata/--no-sqlite-metadata",
+        needs_plugins=[PluginRequirement("rpm", specifier="<=3.24.0")],
+        default=None,
+    ),
```

Comment:
Can you please add a similar guard in the context object, so other consumers than the cli can benefit?

Reply:
Hey @mdellweg , I'm stuck with this.
What I understand from this is that it will prevent api-level calls to uses sqlite-metadata option in the payload for the desired version requirement. Then, it look like I need to create a `PulpRpmSqliteMetadataContext` (similar to [docs](https://docs.pulpproject.org/pulp_cli/architecture/#version-dependent-code-paths)) in "pulp_glue/rpm/context.py" and somehow add it in the `update` subcommand context, but this process is still very unclear to me. Could you help me out?

How can I relate the restriction defined in the EntityContext with the payload option? What should happend in the EntityContext when requirement is not met? How to/should I test that? And more broadly, what could be the those other consumers?

Reply:
First pulp_glue is a library that was split out from the cli. Other consumers are pulp/squeezer and pulpcore (with the replicate feature).
Since this was an additional check for the same, you can even try to only have it over there (I try to get most of the needs_plugin logic handled inside glue because that's the common denominator).
The context class that is effected seems to be `PulpRpmRepositoryContext`, because `sqlite_metadata` is one of it's attributes. The method for such hooks is called `preprocess_entity`.
https://docs.pulpproject.org/pulp_cli/glue_reference/common_context/#pulp_glue.common.context.PulpEntityContext.preprocess_entity

Reply:
Is there a use case where `pulp_option` could be preferred? I found this a bit confusing.
If there is, we could add a section in docs clarifying that.
And in any case, it would be helpfull to add a small example of using the `preprocess_entity` (as I've used here) for adding version constrains to options (maybe [here](https://docs.pulpproject.org/pulp_cli/architecture/#version-dependent-code-paths)).

Reply:
> Is there a use case where `pulp_option` could be preferred? I found this a bit confusing. If there is, we could add a section in docs clarifying that.

`pulp_option` provides a superset of features over `click.option`. Since this project, as many others, grew organically, it is not quite used consistently in all places. To use things like `needs_plugin`, a `pulp_option` is absolutely necessary. I'm still on the fence whether we should recommend using it in (almost; `--config` is an exception for some reason as far as i remember) all places.

> And in any case, it would be helpfull to add a small example of using the `preprocess_entity` (as I've used here) for adding version constrains to options (maybe [here](https://docs.pulpproject.org/pulp_cli/architecture/#version-dependent-code-paths)).

Don't make me stop you. XD

Reply:
OK, sorry, `needs_plugin` does not work with `click.option`, but it still works when the layer below is raising the right exception (Hooray to stack unwinding exception handling!).
Declaring the need at the option level was what we were doing before realizing that pulp_glue had it's independent value as a standalone library. For the sake of the cli it almost does not matter where that exception is raised. But for uses of glue outside of the cli, it does.
I currently believe we should try as best as we can to move (or in your case, add) all version dependency in the glue library so all associated projects can benefit. This also makes the glue library the comprehensive guide to all known Pulp API subtleties.
The question of `pulp_option` over `click.option` is therefor in many cases one of best practice and least surprise. I think i'll vote for "use `pulp_option`" because this make `pulpcore.cli.common.generic` the library to the pulp-cli plugins and we have some control over subcommands that we loose when they use `click.option` directly.

Reply:
Thanks for the extra clarification. I'll open another issue for this.

Reply:
Please don't hesitate to capture any of this in the docs.